pkg/virt-controller/watch/vm.go

```diff
@@ -664,6 +665,108 @@ func (c *VMController) handleCPUChangeRequest(vm *virtv1.VirtualMachine, vmi *vi
 	return nil
 }
 
+func (c *VMController) VMINodeSelectorPatch(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+	var ops []string
+
+	if vm.Spec.Template.Spec.NodeSelector != nil {
+		vmNodeSelector := make(map[string]string)
+		// copy the node selector map
+		for k, v := range vm.Spec.Template.Spec.NodeSelector {
+			vmNodeSelector[k] = v
+		}
+		vmNodeSelectorJson, err := json.Marshal(vmNodeSelector)
+		if err != nil {
+			return err
+		}
+		if vmi.Spec.NodeSelector == nil {
+			ops = append(ops, fmt.Sprintf(`{ "op": "add", "path": "/spec/nodeSelector", "value": %s }`, string(vmNodeSelectorJson)))
+		} else {
+			currentVMINodeSelector, err := json.Marshal(vmi.Spec.NodeSelector)
+			if err != nil {
+				return err
+			}
+			ops = append(ops, fmt.Sprintf(`{ "op": "test", "path": "/spec/nodeSelector", "value": %s }`, string(currentVMINodeSelector)))
+			ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/spec/nodeSelector", "value": %s }`, string(vmNodeSelectorJson)))
+		}
+
+	} else {
+		ops = append(ops, fmt.Sprintf(`{ "op": "remove", "path": "/spec/nodeSelector" }`))
+	}
+	generatedPatch := controller.GeneratePatchBytes(ops)
+
+	_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, generatedPatch, &v1.PatchOptions{})
+
+	return err
+}
+
+func (c *VMController) VMIAffinityPatch(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+	var ops []string
+
+	if vm.Spec.Template.Spec.Affinity != nil {
+		vmAffinity := vm.Spec.Template.Spec.Affinity.DeepCopy()
+		vmAffinityJson, err := json.Marshal(vmAffinity)
+		if err != nil {
+			return err
+		}
+		if vmi.Spec.Affinity == nil {
+			ops = append(ops, fmt.Sprintf(`{ "op": "add", "path": "/spec/affinity", "value": %s }`, string(vmAffinityJson)))
+		} else {
+			currentVMIAffinity, err := json.Marshal(vmi.Spec.Affinity)
+			if err != nil {
+				return err
+			}
+			ops = append(ops, fmt.Sprintf(`{ "op": "test", "path": "/spec/affinity", "value": %s }`, string(currentVMIAffinity)))
+			ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/spec/affinity", "value": %s }`, string(vmAffinityJson)))
+		}
+
+	} else {
+		ops = append(ops, fmt.Sprintf(`{ "op": "remove", "path": "/spec/affinity" }`))
+	}
+
+	_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, controller.GeneratePatchBytes(ops), &v1.PatchOptions{})
+
+	return err
+}
+
+func (c *VMController) handleAffinityChangeRequest(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+	if vmi == nil || vmi.DeletionTimestamp != nil {
+		return nil
+	}
+
+	if vm.Spec.LiveUpdateFeatures == nil || vm.Spec.LiveUpdateFeatures.Affinity == nil {
+		return nil
+	}
+
+	hasNodeSelectorChanged := false
+	if !equality.Semantic.DeepEqual(vm.Spec.Template.Spec.NodeSelector, vmi.Spec.NodeSelector) {
+		hasNodeSelectorChanged = true
+	}
+
+	hasNodeAffinityChanged := false
```

Comment:
supernit - inline

Reply:
@lyarwood Hey. Can you please elaborate on the way to do it inline?

Reply:
```
hasNodeAffinityChanged := !equality.Semantic.DeepEqual(vm.Spec.Template.Spec.Affinity, vmi.Spec.Affinity)
```